feat: implement sabr streaming

FineFindus opened the pull request:
Implements support for streaming using SABR. This is achieved by leveraging a [native rust binary](https://github.com/FineFindus/sabr) that implements the fetching, parsing the UMP stream and extracting the actual media data.

The integration with exoplayer is still very early (read: it does not work yet), though Rust implementation is fully-capable. Help writing the exoplayer integration is appreciated :)

### Open issues
- [ ] due to the reliance on Tokio, the ends up being quite large (~100 MB for debug and currently ~5 for release)
- [ ] this requires additional patches to NewPipeExtractor to extract the necessary metadata (I would like to avoid carrying these forward in our fork)

Closes: https://github.com/libre-tube/LibreTube/issues/8254

FineFindus committed:
feat: retrieve SABR streaming data from extractor

Jeidnx commented:
This looks very interesting. I am particularly interested in the rust side, it seems the repository for that is private though?

FineFindus committed:
chore: prefer SABR when available

FineFindus committed:
feat(sabr): add protobuf files

Adds the necessary infrastructure and files to generate the protobuf classes
used for SABR streaming. This includes a dependency on protoc (the protobuf
compiler), as well as the libraries for Java/Kotlin protobuf interaction.

FineFindus committed:
feat(sabr): add UMP parser

FineFindus committed:
feat(sabr): add SABR client

Ports the SABR streaming client from Rust to Kotlin.
This avoids a round-trip to the native layer for each request, as well as the
additional caching layer, at the cost of worse memory-management and usage.

FineFindus committed:
fix(sabr): correctly set null xtags

FineFindus committed:
fix(sabr): correctly retrieve new media segments

FineFindus committed:
refactor(sabr): use verbose instead of debug logging

FineFindus committed:
feat(sabr): add Representation

FineFindus committed:
feat(sabr): add AdaptationSet

FineFindus committed:
feat(sabr): add manifest

FineFindus committed:
refactor(sabr): convert sabr client to singleton object

This helps to ensure that there is ever only one instance of the client,
avoiding concurrency issues when accessing the object from multiple threads,
which the player does.

FineFindus committed:
refactor(sabr): use representations to retrieve media data

FineFindus committed:
feat(sabr): store and re-use sabr exception

FineFindus committed:
feat(sabr): support retrieving DRC audio

FineFindus committed:
feat(sabr): implement data source

The data source is responsible for ultimately loading the media data from the
server. This is done by requesting the data from the sabr client and then
allowing the player to read the returned data.

FineFindus committed:
refactor(sabr): port to Kotlin

FineFindus committed:
refactor(sabr): remove JVM field tags

FineFindus committed:
fixup unused constructor param

FineFindus committed:
refactor(sabr): remove unused constructors

FineFindus committed:
feat(sabr): implement support for poTokens

Implements support for generating and sending poTokens using the already
existing `PoTokenGenerator`.
A poToken is generator upon requesting the initial stream and may optionally be
regenerated when the server warns that attestation will soon be required.

FineFindus committed:
chore(proto): rename `lmt` to `last_modified` time

FineFindus committed:
feat(sabr): only log unhandled UMP parts

FineFindus committed:
feat(sabr): build and send BufferedRanges

Builds and sends BufferedRanges to the server, notifying it what segment ranges
the client currently has buffered. This influences what segments/quality the
server sends.

FineFindus committed:
fix: remove unused import

FineFindus committed:
refactor(sabr): extract parsing and handling of xtags

FineFindus committed:
feat(sabr): handle formats supporting voice boost

FineFindus committed:
refactor(sabr/xtags): remove unnecessary field

FineFindus committed:
refactor(sabr): mark Representation formatId as non-null

FineFindus committed:
fix(sabr): disable SABR for live-streams

Disables using SABR if the media is currently being live-streamed, as the
current implementation does not support live-streams and the server still sends
DASH and HLS manifests.

FineFindus committed:
refactor(sabr): mark `RepresentationHolder` as data class

FineFindus committed:
refactor(sabr): only handle a single segment per chunk

Instead of handling multiple segments for each chunk, we only handle a single
segment. This avoids another memcopy of the data into a `ByteBuffer`.

FineFindus committed:
refactor(sabr): use idiomatic continue on null selection

FineFindus committed:
style: reformat code

FineFindus committed:
feat(sabr): allow discontinuous playback

Implements support for discontinuous playback, i.e. starting the media at
different time or seeking during playback.

FineFindus committed:
feat(sabr): synchronize buffered segments with chunksource queue

FineFindus committed:
feat(sabr): store init segment

FineFindus committed:
fix(sabr): use buffered and downloaded segments to build buffered ranges

FineFindus committed:
feat(sabr): assert that we never load segments past the end of stream

FineFindus committed:
fix(sabr): use segment start time to request

The server only provides a certain number of segments in advance, so instead of
using the playback position, we pretend we're directly at the segment boundary
when requesting new segments. This may have some drawbacks, as the server may
send lower-quality segments.

FineFindus committed:
feat(sabr): set `Origin` and `Referer` header

FineFindus committed:
refactor(sabr): remove unnecessary null-safety assertion

FineFindus committed:
feat(sabr): set `elapsedWallTimeMs`

Notifies the server about the time in milliseconds that has passed between two
requests.

FineFindus committed:
feat(sabr): avoid retrieving segments from buffered segments

FineFindus committed:
style: reformat code

FineFindus committed:
refactor(sabr): extract common format selection code

FineFindus committed:
refactor(sabr): remove unused `poToken` from manifest

The `poToken` is generated by the SabrClient, when needed.

FineFindus committed:
feat(sabr): inform server about last seek timestamp

FineFindus committed:
refactor(sabr): cleanup

FineFindus committed:
fix(sabr): correctly set xtags from extractor

Fixes an issue with playback of some formats, as the extracted xtags where not
passed from the extractor to the sabr client. This caused some streams to fail
with `sabr.no_audio_selected (0)` as the client did not set the correct
properties.

FineFindus committed:
refactor(sabr): create `FormatId` for `Representation` directly from stream

FineFindus committed:
refactor(sabr): use idiomatic null check

FineFindus committed:
feat(sabr): explicitly handle player reload request

FineFindus committed:
feat(sabr): inform server about manual format selection

FineFindus committed:
refactor(sabr): convert `SabrClient` to a class

Converts the `SabrClient` to be a kotlin class, instead of an object, as
objects cannot be easily destroyed/reset. This may lead to issues, where
old data is send when requesting data for a new stream.

FineFindus committed:
docs(sabr): document Xtags

FineFindus committed:
feat(sabr): informat server about estimated bandwidth

FineFindus committed:
refactor(sabr): use idiomatic Kotlin builder

FineFindus committed:
refactor(sabr): use idiomatic Kotlin nullcheck

FineFindus committed:
feat(sabr): experimentally cache generated PoToken

To avoid generating the same token twice (once for the initial extractor
request and a second time for the SABR requests), try to cache the token when
it's first generated.

FineFindus committed:
refactor(sabr): mark `adaptationSets` setter as private

FineFindus committed:
fix(sabr): use approximate stream duration milliseconds

Fixes an issue, where the stream would end slightly too early, due to
`durationMs` being derived from seconds, truncating the last few milliseconds.

FineFindus committed:
refactor(sabr): remove unsatifiable unbound index check

As the `segmentCount` would always be `>= 0`, it's not possible to take the
value of `-1`, and can thus be removed.

FineFindus committed:
fix(sabr): use correct segment number for media segments

The `chunkIndex` does not count the init segment, leading to the last segment
(i.e. `chunkIndex.length`) not being request/played.

FineFindus committed:
fix(sabr): avoid unnecessary request for init segment

The player may request the init segment multiple times, so we check if
it exists, before requesting it from the server again, thus avoiding an
unnecessary request.

FineFindus committed:
fix(sabr): avoid storing media data for buffered segments

Instead of storing the full media data for all buffered segments, we only store
the metadata of the segments. This avoids out-of-memory errors.

FineFindus committed:
fix(sabr): clear unbuffered segments when requesting new segments

Removes all segments, that have now been requested by the player yet, when
requesting new segments from the server. This should avoid segments being
forgotten, as they're never requested (e.g. due to seeking) and thus
effectively being leaked.

FineFindus committed:
fix(sabr): clear previous format when switching formats

Clears the data of the previously selected format, when selecting a
new format. This avoid advertising the stale formats as selected/buffered to
the server, causing it to not send these segments for the new format.
This fixes playback issues for short(er) videos.

FineFindus committed:
fix(SABR): update selected track when track selection changes

FineFindus committed:
feat(SABR): skip re-selecting already selected format

FineFindus committed:
fix(SABR): only update selected format for the next request

Only updates the selected format for the next media request, ensuring that some
media data is already buffered from the current request. This avoids playback
errors, when only one stream would load.

FineFindus committed:
refactor(SABR): directly return server response payload

FineFindus committed:
feat(SABR): clear formats on media refresh

FineFindus committed:
feat(SABR): support external subtitles

Implements support for using external subtitles (provided by an additional
stream), by using separate media sources for the subtitles and merging them for
playback.

FineFindus committed:
feat(SABR): enable lazy-loading of subtitle streams

Enables lazy-loading for the subtitle streams, as the format of the subtitles
is already known. As the method to enable lazy-loading is private, it is
accessed through reflection.

FineFindus committed:
refactor(SABR/Xtags): keep attributes with non-integer values

FineFindus committed:
feat(SABR): extract language from Xtags

FineFindus committed:
style(SABR): fix typo

FineFindus committed:
refactor(SABR): remove unused variables

FineFindus committed:
style(SABR): remove solved TODOs

FineFindus committed:
build: use NewPipeExtractor with SABR support
